### PR TITLE
PiccoloCAN parameters

### PR DESCRIPTION
**Pull request by SchrodingersGat**
This PR follows on from the work in https://github.com/ArduPilot/ardupilot/pull/12947.

I have added some parameters to control the operation of the PiccoloCAN interface. At this stage it still only supports ESC (servos coming soon, though!)

This PR adds two ESC-specific parameters:

**CAN_Dx_PC_ESC_MS** - *ESC Command Rate*

This parameter controls the command output rate (in ms) for command packets sent out to any Piccolo-compatible ESCs on the bus.

**CAN_Dx_PC_ESC_BM** - *ESC Selection Mask*

Determines which ESC channels are actually sent out on the bus. It may be that some ESC are on the PiccoloCAN bus and some are PWM, or maybe the user only wants to run some ESCs for testing. This selection mask controls which ESC channels are used for:

- Control output
- Arm checking

If any ESC are present which are not selected here, their data will still be logged and sent over mavlink.

### CODE REVIEW

**Comment by SchrodingersGat** [comment]
@tridge any thoughts on this one? :)

**Comment by SchrodingersGat** [comment]
Friendly bump on this one - any chance this can get some attention?

**Comment by auturgy** [comment]
just dropping a devcall tag on this so it gets looked at.

File: libraries/AP_BoardConfig/AP_BoardConfig_CAN.cpp [hunk 1]
```diff
@@ -181,9 +181,11 @@ void AP_BoardConfig_CAN::init()
                 _drivers[i]._driver = _drivers[i]._pcan = new AP_PiccoloCAN;
 
                 if (_drivers[i]._driver == nullptr) {
-                    AP_BoardConfig::config_error("PiccoloCAN init failed");
+                    AP_HAL::panic("Failed to allocate PiccoloCAN %d\n\r", i + 1);
```
Comment (tridge) [inline libraries/AP_BoardConfig/AP_BoardConfig_CAN.cpp:184]: why the change to a panic?

File: libraries/AP_BoardConfig/canbus_driver.cpp [hunk 2]
```diff
@@ -34,7 +34,7 @@ const AP_Param::GroupInfo AP_BoardConfig_CAN::Driver::var_info[] = {
     // @DisplayName: Enable use of specific protocol over virtual driver
     // @Description: Enabling this option starts selected protocol that will use this virtual driver
     // @Values{Copter,Plane,Sub}: 0:Disabled,1:UAVCAN,2:KDECAN,3:ToshibaCAN,4:PiccoloCAN
-    // @Values: 0:Disabled,1:UAVCAN,3:ToshibaCAN,4:PiccoloCAN
+    // @Values: 0:Disabled,1:UAVCAN,2:KDECAN,3:ToshibaCAN,4:PiccoloCAN
```
Comment (tridge) [inline libraries/AP_BoardConfig/canbus_driver.cpp:37]: why this change?

Comment (peterbarker) [inline libraries/AP_BoardConfig/canbus_driver.cpp:37]: Could use `@Values{Plane, copter, Sub}: ` as another line instead of making the change here.

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp [hunk 3]
```diff
@@ -43,12 +45,48 @@
...
+    AP_GROUPINFO("ESC_BM", 1, AP_PiccoloCAN, _esc_bm, 0),
+
+    // @Param: ESC_MS
+    // @DisplayName: ESC command rate
+    // @Description: Output rate of ESC command messages
+    // @Units: ms
+    // @Range: 2 100
+    AP_GROUPINFO("ESC_MS", 2, AP_PiccoloCAN, _esc_ms, PICCOLO_MSG_RATE_MS_DEFAULT),
+
+    // @Param: SRV_BM
```
Comment (peterbarker) [inline libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp:65]: Maybe only add servo parameters once we support them?

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp [hunk 4]
```diff
@@ -362,7 +409,9 @@ void AP_PiccoloCAN::send_esc_messages(void)
     // TODO - How to buffer CAN messages properly?
     // Sending more than 2 messages at each loop instance means that sometimes messages are dropped
 
-    if (hal.util->get_soft_armed()) {
+    // No ESCs are selected? Don't send anything
+    if (_esc_bm == 0x00) {
```
Comment (rmackay9) [inline libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp:413]: It's a bit funny to have an "if" with nothing within the curly braces.  Maybe just make it a !.

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp [hunk 5]
```diff
@@ -136,13 +175,22 @@ void AP_PiccoloCAN::loop()
...
+        }
+
+        // Transmit servo commands at regular intervals
+        if (srv_tx_counter++ > _srv_ms) {
+            srv_tx_counter = 0;
 
-            // Transmit ESC commands
-            send_esc_messages();
+            if (_srv_ms >= PICCOLO_MSG_RATE_MS_MIN) {
+                // TODO - Transmit servo messages
```
Comment (tridge) [inline libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp:192]: shouldn't add the params till we can use them

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp [hunk 6]
```diff
@@ -43,12 +45,48 @@
...
 
+// table of user-configurable Piccolo CAN bus parameters
+const AP_Param::GroupInfo AP_PiccoloCAN::var_info[] = {
+
+    // @Param: ESC_BM
+    // @DisplayName: ESC channels
+    // @Description: Bitmask defining which ESC (motor) channels are to be transmitted over Piccolo CAN
+    // @Bitmask: 0: ESC 1, 1: ESC 2, 2: ESC 3, 3: ESC 4, 4: ESC 5, 5: ESC 6, 6: ESC 7, 7: ESC 8, 8: ESC 9, 9: ESC 10, 10: ESC 11, 11: ESC 12, 12: ESC 13, 13: ESC 14, 14: ESC 15, 15: ESC 16
+    // @User: Advanced
+    AP_GROUPINFO("ESC_BM", 1, AP_PiccoloCAN, _esc_bm, 0),
```
Comment (rmackay9) [inline libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp:56]: How about we get rid of this ESC_BM bitmask and just assume that all SERVOx_FUNCTION = motorX outputs are PiccoloCAN?

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp [hunk 7]
```diff
@@ -43,12 +45,48 @@
...
+    // @Units: ms
+    // @Range: 2 100
+    AP_GROUPINFO("ESC_MS", 2, AP_PiccoloCAN, _esc_ms, PICCOLO_MSG_RATE_MS_DEFAULT),
+
+    // @Param: SRV_BM
+    // @DisplayName: Servo channels
+    // @Description: Bitmask defining which servo channels are to be transmitted over PiccoloCAN
+    // @Bitmask: 0: Servo 1, 1: Servo 2, 2: Servo 3, 3: Servo 4, 4: Servo 5, 5: Servo 6, 6: Servo 7, 7: Servo 8, 8: Servo 9, 9: Servo 10, 10: Servo 11, 11: Servo 12, 12: Servo 13, 13: Servo 14, 14: Servo 15, 15: Servo 16
+    // @User: Advanced
+    AP_GROUPINFO("SRV_BM", 3, AP_PiccoloCAN, _srv_bm, 0),
```
Comment (rmackay9) [inline libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp:70]: As @peterbarker suggests, let's remove this parameter if it's not used.

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp [hunk 8]
```diff
@@ -43,12 +45,48 @@
...
+
+    // @Param: SRV_MS
+    // @DisplayName: Servo command rate
+    // @Description: Output rate of servo command messages
+    // @Units: ms
+    // @Range: 2 100
+    AP_GROUPINFO("SRV_MS", 4, AP_PiccoloCAN, _srv_ms, PICCOLO_MSG_RATE_MS_DEFAULT),
```
Comment (rmackay9) [inline libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp:77]: let's remove this unused param too

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp [hunk 9]
```diff
@@ -43,12 +45,48 @@
...
+    // @Bitmask: 0: ESC 1, 1: ESC 2, 2: ESC 3, 3: ESC 4, 4: ESC 5, 5: ESC 6, 6: ESC 7, 7: ESC 8, 8: ESC 9, 9: ESC 10, 10: ESC 11, 11: ESC 12, 12: ESC 13, 13: ESC 14, 14: ESC 15, 15: ESC 16
+    // @User: Advanced
+    AP_GROUPINFO("ESC_BM", 1, AP_PiccoloCAN, _esc_bm, 0),
+
+    // @Param: ESC_MS
+    // @DisplayName: ESC command rate
+    // @Description: Output rate of ESC command messages
+    // @Units: ms
+    // @Range: 2 100
+    AP_GROUPINFO("ESC_MS", 2, AP_PiccoloCAN, _esc_ms, PICCOLO_MSG_RATE_MS_DEFAULT),
```
Comment (tridge) [inline libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp:63]: could this be a RT as a rate in Herz? That would match uavcan params

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp (hunk 8 above)
Comment (tridge) [inline libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp:77]: can this be rate in Hz?

**Comment by SchrodingersGat** [comment]
Thanks for the feedback all - I'll start working through this and ping you when I think it's ready.

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp (hunk 8 above)
Comment (SchrodingersGat) [reply libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp]: The parameter "SRV_MS" has now been removed, pending implementation of CAN servo support.

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp (hunk 9 above)
Comment (SchrodingersGat) [reply libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp]: `ESC_MS` is now `ESC_RT` and is the transmission rate in Hz

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp (hunk 8 above)
Comment (SchrodingersGat) [reply libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp]: This unused param has now been removed as requested

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp (hunk 7 above)
Comment (SchrodingersGat) [reply libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp]: `SRV_BM` parameter has now been removed

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp (hunk 5 above)
Comment (SchrodingersGat) [reply libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp]: This has now been removed from this PR

File: libraries/AP_BoardConfig/canbus_driver.cpp (hunk 2 above)
Comment (SchrodingersGat) [reply libraries/AP_BoardConfig/canbus_driver.cpp]: This was polluted from my local repo - I have now reverted this in this PR

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp (hunk 3 above)
Comment (SchrodingersGat) [reply libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp]: This parameter is now removed from this PR

File: libraries/AP_BoardConfig/AP_BoardConfig_CAN.cpp (hunk 1 above)
Comment (SchrodingersGat) [reply libraries/AP_BoardConfig/AP_BoardConfig_CAN.cpp:184]: Only to make it similar to the call a few lines up, for uavcan driver

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp (hunk 4 above)
Comment (SchrodingersGat) [reply libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp:398]: I've adjusted the logic here to make it more readable

**Comment by SchrodingersGat** [comment]
I believe it has all been addressed now - please LMK if there are any other points of contention.

**Comment by SchrodingersGat** [comment]
@tridge when you have some time, can I please have another quick review of this one?

**Comment by SchrodingersGat** [comment]
Hi any chance this could be addressed if someone has some time? 👍

**Comment by tridge** [comment]
@SchrodingersGat I'm really sorry we dropped the ball on this. As you may know, we've done a major overhaul of the CAN system layer recently. Any chance you could rebase? Otherwise I could try a rebase if you can test.

**Comment by SchrodingersGat** [comment]
I've merged master back in, is this enough? Or do you need me to rebase the changes off latest master instead?

**Comment by tridge** [comment]
> I've merged master back in, is this enough? Or do you need me to rebase the changes off latest master instead?

it does need to be a rebase, sorry. I'm happy to do it if you'd like me to

**Comment by SchrodingersGat** [comment]
@tridge I think I've done this correctly...

It took some figuring out to make the parameter changed work with the new CAN framework, but it is now working well on my end.

Please LMK if there is anything else you need me to do. Cheers.

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp [hunk 10]
```diff
@@ -112,19 +135,29 @@ void AP_PiccoloCAN::loop()
...
     uint16_t frame_id_device;   // Device identifier
 
     uint64_t timeout;
 
+    uint16_t escCmdRateMs;
+
     while (true) {
 
+        // Calculate the output rate (in ms) based on the configured rate (in Hz)
+        if (_esc_hz < PICCOLO_MSG_RATE_HZ_MIN) {
```
Comment (tridge) [inline libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp:151]: this should use constrain_int16()

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.h [hunk 11]
```diff
@@ -125,6 +136,10 @@ class AP_PiccoloCAN : public AP_CANDriver
 
     } _esc_info[PICCOLO_CAN_MAX_NUM_ESC];
 
+    // Piccolo CAN parameters
+    AP_Int32 _esc_bm;       //! ESC selection bitmask
+    AP_Int32 _esc_hz;       //! ESC update rate (Hz)
```
Comment (tridge) [inline libraries/AP_PiccoloCAN/AP_PiccoloCAN.h:141]: I think 16 bit will do for the rate in Hz ?

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by tridge** [comment]
@SchrodingersGat thanks for the update! only found 2 very minor things.

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp (hunk 10 above)
Comment (SchrodingersGat) [reply libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp]: Done

File: libraries/AP_PiccoloCAN/AP_PiccoloCAN.h (hunk 11 above)
Comment (SchrodingersGat) [reply libraries/AP_PiccoloCAN/AP_PiccoloCAN.h]: Done

**Comment by SchrodingersGat** [comment]
> @SchrodingersGat thanks for the update! only found 2 very minor things.

Thanks, these have been addressed :)

**Comment by SchrodingersGat** [comment]
Hi @tridge any more requirements for this? :)

**Review by tridge — APPROVED** [review]
